lib/src/default_index_store.rs

```diff
@@ -266,3 +274,2390 @@ fn topo_order_earlier_first(
+            [&id_2, &id_1].map(Clone::clone)
+        );
+        assert_eq!(
+            walk_commit_ids(&[&id_7, &id_8].map(Clone::clone), &[], 0..2),
+            [&id_8, &id_7, &id_4].map(Clone::clone)
+        );
+        assert_eq!(
+            walk_commit_ids(&[&id_6, &id_7].map(Clone::clone), &[], 0..3),
+            [&id_7, &id_6, &id_5, &id_4, &id_3, &id_2].map(Clone::clone)
+        );
+        assert_eq!(
+            walk_commit_ids(&[&id_6, &id_7].map(Clone::clone), &[], 2..3),
+            [&id_4, &id_3, &id_2].map(Clone::clone)
+        );
+
+        // Ancestors of both wanted and unwanted commits are not walked
+        assert_eq!(
+            walk_commit_ids(&[&id_5].map(Clone::clone), &[&id_2].map(Clone::clone), 1..5),
+            [&id_4, &id_3].map(Clone::clone)
+        );
+    }
+
+    #[test]
+    fn test_heads() {
+        let mut new_change_id = change_id_generator();
+        let mut index = MutableIndex::full(3, 16);
+        // 5
+        // |\
+        // 4 | 3
+        // | |/
+        // 1 2
+        // |/
+        // 0
+        let id_0 = CommitId::from_hex("000000");
+        let id_1 = CommitId::from_hex("111111");
+        let id_2 = CommitId::from_hex("222222");
+        let id_3 = CommitId::from_hex("333333");
+        let id_4 = CommitId::from_hex("444444");
+        let id_5 = CommitId::from_hex("555555");
+        index.add_commit_data(id_0.clone(), new_change_id(), &[]);
+        index.add_commit_data(id_1.clone(), new_change_id(), &[id_0.clone()]);
+        index.add_commit_data(id_2.clone(), new_change_id(), &[id_0.clone()]);
+        index.add_commit_data(id_3.clone(), new_change_id(), &[id_2.clone()]);
+        index.add_commit_data(id_4.clone(), new_change_id(), &[id_1.clone()]);
+        index.add_commit_data(id_5.clone(), new_change_id(), &[id_4.clone(), id_2.clone()]);
+
+        // Empty input
+        assert!(index.heads(&mut [].iter()).is_empty());
+        // Single head
+        assert_eq!(index.heads(&mut [id_4.clone()].iter()), vec![id_4.clone()]);
+        // Single head and parent
+        assert_eq!(
+            index.heads(&mut [id_4.clone(), id_1].iter()),
+            vec![id_4.clone()]
+        );
+        // Single head and grand-parent
+        assert_eq!(
+            index.heads(&mut [id_4.clone(), id_0].iter()),
+            vec![id_4.clone()]
+        );
+        // Multiple heads
+        assert_eq!(
+            index.heads(&mut [id_4.clone(), id_3.clone()].iter()),
+            vec![id_3.clone(), id_4]
+        );
+        // Merge commit and ancestors
+        assert_eq!(
+            index.heads(&mut [id_5.clone(), id_2].iter()),
+            vec![id_5.clone()]
+        );
+        // Merge commit and other commit
+        assert_eq!(
+            index.heads(&mut [id_5.clone(), id_3.clone()].iter()),
+            vec![id_3, id_5]
+        );
+    }
+
+    #[test]
+    fn test_hex_prefix_prefixes() {
```

Comment:
nit: these two tests don't depend on the index impl, so better to not move.

Reply:
Good catch. Fixed.